### PR TITLE
V2 fix render 813

### PR DESCRIPTION
**Pull request by hrgdavor**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

I tried to run
standardx .\packages\web\src\ui\views\viewer.js
and 
npm run lint

but I get an error 
Error: Failed to load config "standard" to extend from.

looks like newer version of standardx require plugins to be local

I have added eslint-config-standard as dev dependency and after that it worked

### CODE REVIEW

**Comment by z3dev** [comment]
> I tried to run
> standardx .\packages\web\src\ui\views\viewer.js
> and
> npm run lint
> but I get an error

Correct. Lint has to be run from the main project, OpenJSCAD.org. The lint command can be changed to lint a single directory.

**Review by z3dev — CHANGES_REQUESTED** [review]

File: packages/web/src/ui/views/viewer.js [hunk 1]
```diff
@@ -86,44 +91,71 @@ const viewer = (state, i18n) => {
...
         zoomDelta = 0
+        changed = true
       }
+      return changed
     }
 
     // the heart of rendering, as themes, controls, etc change
     const updateAndRender = (timestamp) => {
-      doRotatePanZoom()
+      if (doRotatePanZoom() || controls.autoRotate.enabled) {
```
[the remark this answers is not in the document]
Comment (hrgdavor) [reply packages/web/src/ui/views/viewer.js:122]: but doRotatePanZoom has to be called from somewhere, is it triggered from somwhere else ?

File: packages/web/src/ui/views/viewer.js [hunk 2]
```diff
@@ -86,44 +91,71 @@ const viewer = (state, i18n) => {
...
+        renderIndicator.innerHTML = 'R'
+        const style = renderIndicator.style
+        style.position = 'absolute'
+        style.bottom = '10px'
+        style.right = '10px'
+        document.body.renderIndicator = renderIndicator
+        document.body.appendChild(renderIndicator)
+      }
+
+      if (!renderUntil || renderUntil > Date.now()) {
```
[the remark this answers is not in the document]
Comment (hrgdavor) [reply packages/web/src/ui/views/viewer.js:143]: orbit controls have some elasticity, so this leaves some time for render to continue to have like a time-buffer that lets orbit animation to finish.

**Review by z3dev — CHANGES_REQUESTED** [review]
Small issue with double clicking (automatic zoom). Please review.

Performance is really nice! 👍 

Before:
```
PID   COMMAND      %CPU  TIME     #TH   #WQ  #PORTS MEM    PURG   CMPRS  PGRP PPID STATE    BOOSTS          %CPU_ME %CPU_OTHRS UID  FAULTS    COW
530   firefox      42.7  04:21:38 97/1  3    13116  1283M- 44M    0B     530  1    running  *0[9183]        0.00000 0.00000    501  48846162+ 15524
9704  plugin-conta 33.8  00:18.01 43    1    191    150M+  0B     0B     530  530  sleeping *0[2]           0.00000 0.00000    501  92453+    1741
```
After:
```
530   firefox      2.1   04:22:05 95    1    13145  1276M  44M    0B     530  1    sleeping *0[9228]        0.00000 0.00000    501  49006865+ 15525
9707  plugin-conta 1.8   00:05.11 40    1    195    99M+   8192B  0B     530  530  sleeping *0[2]           0.00000 0.00000    501  83532+    1741
```

File: packages/web/src/ui/views/viewer.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/web/src/ui/views/viewer.js:131]: > orbit controls have some elasticity, so this leaves some time for render to continue to have like a time-buffer that lets orbit animation to finish.

I tried both this version versus previous version. I didn't find any differences when resizing the screen, rotating shapes, or auto-rotation. It seemed to make Chrome slightly 'delayed' when resizing the screen, but Firefox and Safari didn't change.

I'm using MacOS for development. Should I try Windows? Linux?

[the remark this answers is not in the document]
Comment (hrgdavor) [reply packages/web/src/ui/views/viewer.js]: I am testing on windows. I am confident it is sufficient we tested at least two OS. I did not notice resize difference between openjscad.xyz and this version.

File: packages/web/src/ui/views/viewer.js [hunk 3]
```diff
@@ -86,44 +92,55 @@ const viewer = (state, i18n) => {
       })
```
[the remark this answers is not in the document]
Comment (hrgdavor) [reply packages/web/src/ui/views/viewer.js:93]: fixed

File: packages/web/src/ui/views/viewer.js [hunk 4]
```diff
@@ -83,47 +89,60 @@ const viewer = (state, i18n) => {
...
-      doRotatePanZoom()
+      if (doRotatePanZoom() || controls.autoRotate.enabled) {
+        moveRender()
+      }
 
-      const updated = orbitControls.update({ controls, camera })
-      controls = { ...controls, ...updated.controls }
-      camera.position = updated.camera.position
-      perspectiveCamera.update(camera)
+      if (!renderUntil || renderUntil > Date.now()) {
```
Comment (z3dev) [inline packages/web/src/ui/views/viewer.js:128]: Just one more really small request... sorry. Set...
`let renderUntil = Number.MAX_VALUE`
And then this IF only needs...
`if (renderUntil > Date.now())`

**Review by z3dev — CHANGES_REQUESTED** [review]
Sorry. I have a little change.

File: packages/web/src/ui/views/viewer.js (hunk 4 above)
Comment (hrgdavor) [reply packages/web/src/ui/views/viewer.js:128]: np, it is a nice hint. Thnx!

**Review by z3dev — APPROVED** [review]
OK. These changes look good. Thanks again.

**Pull request merged**